### PR TITLE
Make "ignored..." comments in the C++ header generation optional

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
Let the user opt-in if he wants to track whether certain functions are not exposed to the C++ interface.

CC @wilzbach  @kinke

### CODE REVIEW

**Review by wilzbach — COMMENTED** [review]
Nice! Could you maybe add
- overview page of all available options when help is passed (see e.g. preview)
- add an explanation of the available options to cli.d so that it appears in the documentation

**Comment by MoonlightSentinel** [comment]
Sure, this was just an initial implmementation to evaluate this idea.

**Comment by MoonlightSentinel** [comment]
Done. Should i also add a changelog entry?

File: src/dmd/cli.d [hunk 1]
```diff
@@ -369,8 +369,15 @@ dmd -cov -unittest myprog.d
         Option("Hf=<filename>",
             "write 'header' file to filename"
         ),
-        Option("HC",
-            "generate C++ 'header' file"
+        Option("HC=[silent|verbose]",
```
Comment (kinke) [inline src/dmd/cli.d:372]: In general, this DMD style is unintuitive to me. `[...]` would be something optional in my eyes. Something like `HC[=<silent|verbose>]` (I'm sure you get the meaning) would IMO be preferrable, keeping backwards compatibility, but allowing to use the new explicit variants.

**Comment by wilzbach** [comment]
> Done. Should i also add a changelog entry?

You don't have to. As you can see by the broken output of frontend.h the header generation isn't ready to be used in the outside world - and even if we're just removing comments. So I won't object if you add a changelog, but I don't require it either.

File: src/dmd/cli.d [hunk 2]
```diff
@@ -369,8 +369,15 @@ dmd -cov -unittest myprog.d
...
+            "generate C++ 'header' files using the given configuration:",
+            `$(DL
+            $(DT silent)$(DD only list extern(C[++]) declaritions)
```
Comment (wilzbach) [inline src/dmd/cli.d:375]: Declarations

**Review by wilzbach — APPROVED** [review]

File: src/dmd/cli.d (hunk 1 above)
Comment (MoonlightSentinel) [reply src/dmd/cli.d]: Good point, changed it to `-HC[=[silent|verbose]]` based on the help text for `-check`.

**Comment by MoonlightSentinel** [comment]
Added a changelog to get some more attention for the header generator.

File: src/dmd/globals.h [hunk 3]
```diff
@@ -89,6 +89,14 @@ enum CppStdRevision
     CppStdRevisionCpp17 = 201703
 };
 
+/// Configuration for the C++ header generator
+enum CxxHeaderMode
```
Comment (kinke) [inline src/dmd/globals.h:93]: Please use `enum class`. Is the mixed casing of the members intentional?

Comment (MoonlightSentinel) [reply src/dmd/globals.h]: > Please use `enum class`.

Fixed

> Is the mixed casing of the members intentional?

No, just a typo.

**Review by kinke — APPROVED** [review]

**Comment by wilzbach** [comment]
The new CodeCov annotations are pretty cool, lots of code in the header generator that is untested. :/

**Pull request merged**